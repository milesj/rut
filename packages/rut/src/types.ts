/* eslint-disable @typescript-eslint/no-explicit-any, @typescript-eslint/no-namespace */

import React from 'react';
import Element from './Element';

export interface RendererOptions {
  /** Options to pass to the debugger. */
  debugger?: DebugOptions;
  /** Mock a ref found within the current render tree. */
  mockRef?: (element: React.ReactElement) => unknown;
  /** Wraps the root element in `React.StrictMode`. */
  strict?: boolean;
  /** Wraps the root element in the provided React element. */
  wrapper?: React.ReactElement;
}

export interface DebugOptions {
  /** Group props into the following: key & ref, truthy booleans, everything
  else, event handlers. Defaults to `true`. */
  groupProps?: boolean;
  /** Include host elements (DOM) in the output. Defaults to `true`. */
  hostElements?: boolean;
  /** Include `key` and `ref` props in the output. Defaults to `true`. */
  keyAndRef?: boolean;
  /** Max length of arrays and objects before truncating. Defaults to 5. */
  maxLength?: number;
  /** Do not render children. Defaults to `false`. */
  noChildren?: boolean;
  /** Include React elements in the output. Defaults to `true`. */
  reactElements?: boolean;
  /** Do not log to the console and instead return the output. Defaults to
  `false`. */
  return?: boolean;
  /** Sort the props within each grouping from A-Z. Defaults to `true`. */
  sortProps?: boolean;
}

export interface DispatchOptions {
  propagate?: boolean;
}

export type ArgsOf<T> = T extends (...args: infer A) => unknown ? A : never;

export type ReturnOf<T> = T extends (...args: unknown[]) => infer R ? R : unknown;

export type PropsOf<T> = T extends Element<infer P>
  ? P
  : T extends React.ReactElement<infer P>
  ? P
  : T extends React.ComponentType<infer P>
  ? P
  : {};

export type StructureOf<T> = { [K in keyof T]: T[K] };

export interface UnknownProps {
  [name: string]: unknown;
}

export interface TestNode {
  children: string | (TestNode | string)[];
  instance: {
    context: object;
    props: UnknownProps;
    refs: { [key: string]: any };
    state: object;
  } | null;
  parent: TestNode | null;
  props: UnknownProps;
  type: React.ElementType;
}

export interface FiberNode {
  child: FiberNode | null;
  elementType: React.ElementType;
  index: number;
  key: string | null;
  mode: number;
  ref: any;
  return: FiberNode | null;
  sibling: FiberNode | null;
  tag: number;
  type: React.ElementType;
}

export type AtIndexType = 'first' | 'last' | number;

export type Predicate = (node: TestNode, fiber: FiberNode) => boolean;

export interface MatchResult {
  context?: string;
  message: string;
  notMessage: string;
  passed: boolean;
}

export type NodeType =
  | 'class-component'
  | 'forward-ref'
  | 'function-component'
  | 'host-component'
  | 'memo';

export type HostComponentType = keyof JSX.IntrinsicElements;

export type HostProps<T extends HostComponentType> = JSX.IntrinsicElements[T];

declare module 'react-test-renderer' {
  interface ReactTestInstance {
    _fiber: FiberNode;
  }
}

declare global {
  namespace jest {
    interface Matchers<R> {
      toBeChecked(): R;
      toBeDisabled(): R;
      toBeElementType(type: React.ElementType): R;
      toBeNodeType(type: NodeType): R;
      toContainNode(node: NonNullable<React.ReactNode>): R;
      toHaveClassName(name: string): R;
      toHaveKey(value: string | number): R;
      toHaveProp<K extends keyof PropsOf<R>>(name: K, value?: unknown): R;
      toHaveProps(props: Partial<PropsOf<R>>): R;
      toHaveRendered(): R;
      toHaveValue(value: unknown): R;
    }
  }
}
