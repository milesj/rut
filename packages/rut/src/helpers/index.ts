export * from './predicates';
export * from './utils';
