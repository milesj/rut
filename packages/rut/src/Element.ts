/* eslint-disable @typescript-eslint/no-unused-vars */

import React from 'react';
import { ReactTestInstance } from 'react-test-renderer';
import { getTypeName } from './internals/react';
import { doAct, doAsyncAct } from './internals/act';
import { debug } from './internals/debug';
import { getPropForDispatching } from './internals/element';
import { whereTypeAndProps } from './predicates';
import {
  Predicate,
  DispatchOptions,
  DebugOptions,
  UnknownProps,
  AtIndexType,
  QueryOptions,
  AdapterRendererOptions,
  ElementType,
} from './types';

export default abstract class Element<
  Type extends ElementType = ElementType,
  Props = never,
  Host = unknown
> {
  // @ts-expect-error Set after instantiation
  options: AdapterRendererOptions;

  protected element: ReactTestInstance;

  private readonly isRutElement = true;

  constructor(element: ReactTestInstance) {
    this.element = element;
  }

  /**
   * Log and return a JSX representation of the current *reconciled* React component tree.
   * Does not include exotic components or nodes, such as:
   *
   *  - Context consumers and providers
   *  - Roots, portals, modes
   *  - Profiler, Suspense
   *  - Fragments
   */
  debug = (options: DebugOptions = {}) => debug(this.element, options);

  /**
   * Dispatch an event listener for the defined prop name. Requires a synthetic event
   * based on the original event type.
   *
   * Note: This may only be executed on host components.
   */
  dispatch(name: string, eventOrConfig?: unknown, options: DispatchOptions = {}): this {
    const prop = getPropForDispatching(this, name);
    const event = this.createSyntheticEvent(name, eventOrConfig, this.element.type);

    // istanbul ignore next
    if (options.propagate) {
      // eslint-disable-next-line no-console
      console.warn('Event propagation is experimental and is not fully implemented.');
    }

    doAct(() => prop(event), this.options.applyPatches);

    return this;
  }

  /**
   * Like `dispatch` but also awaits the event so that async calls have time to finish.
   */
  async dispatchAndWait(
    name: string,
    eventOrConfig?: unknown,
    options: DispatchOptions = {},
  ): Promise<void> {
    const prop = getPropForDispatching(this, name);
    const event = this.createSyntheticEvent(name, eventOrConfig, this.element.type);

    // istanbul ignore next
    if (options.propagate) {
      // eslint-disable-next-line no-console
      console.warn('Event propagation is experimental and is not fully implemented.');
    }

    await doAsyncAct(() => prop(event), this.options.applyPatches);
  }

  /**
   * Search through the current tree for all elements that match the defined React
   * component or HTML type. If any are found, a list of `Element`s is returned.
   */
  find(type: ElementType, props?: UnknownProps): Element<ElementType>[] {
    return this.query(whereTypeAndProps(type, props));
  }

  /**
   * Search through the current tree for all elements that match the defined React
   * component or HTML type. If any are found, return the `Element` at the defined
   * index. Accepts shorthand `first` and `last` indices.
   */
  findAt(type: ElementType, at: AtIndexType, props?: UnknownProps): Element<ElementType> {
    const results = this.query(whereTypeAndProps(type, props));
    let index: number;

    if (at === 'first') {
      index = 0;
    } else if (at === 'last') {
      index = results.length - 1;
    } else if (typeof at === 'number') {
      index = at;
    } else {
      throw new TypeError(`Invalid index type "${at}".`);
    }

    const value = results[index];

    if (!value) {
      throw new Error(
        `Expected to find an element at index ${index} for \`${getTypeName(type)}\`.`,
      );
    }

    return value;
  }

  /**
   * Search through the current tree for all elements that match the defined React
   * component or HTML type. If exactly 1 is found, a `Element`s is returned,
   * otherwise an error is thrown.
   */
  findOne(type: ElementType, props?: UnknownProps): Element<ElementType> {
    const results = this.find(type, props);

    if (results.length !== 1) {
      throw new Error(
        `Expected to find 1 element for \`${getTypeName(type)}\`, found ${results.length}.`,
      );
    }

    return results[0];
  }

  /**
   * Return the name of the component, element, or node as JSX.
   */
  name(jsx: boolean = false): string {
    // Use the raw fiber types for names, as they preserve the internal node structures
    const name = getTypeName(this.element._fiber.elementType || this.element._fiber.type);

    return jsx ? `<${name} />` : name;
  }

  /**
   * Search through the current child tree using a custom predicate, which is passed the
   * ReactTestRenderer node and internal React fiber node. If any are found,
   * a list of `Element`s is returned.
   */
  query<T extends ElementType>(predicate: Predicate, options?: QueryOptions): Element<T>[] {
    return this.element
      .findAll((node) => predicate(node, node._fiber), { deep: true, ...options })
      .map((node) => {
        const element = this.options.createElement(node);

        element.options = this.options;

        return element;
      });
  }

  /**
   * Return a ref associated with the current component.
   * If a ref is stored on the internal fiber node, it will be returned.
   * Otherwise an instance, callback, or string ref may be referenced by name.
   */
  ref<T>(name?: string): T | null {
    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
    const inst = this.element.instance;
    const fiber = this.element._fiber;

    if (!name || fiber.ref) {
      return fiber.ref;
    }

    // Callback refs
    if (name in inst) {
      return inst[name];
    }

    // String refs
    if (inst.refs[name]) {
      return inst.refs[name];
    }

    return null;
  }

  /**
   * Return name when cast as a string.
   */
  toString(): string {
    return this.name(true);
  }

  /**
   * Create a synthetic event specific to this renderer.
   */
  abstract createSyntheticEvent(
    name: string,
    eventOrConfig: unknown,
    type: ElementType,
  ): React.SyntheticEvent;
}
