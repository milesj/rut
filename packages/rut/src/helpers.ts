/* eslint-disable complexity */

import React from 'react';
import * as ReactIs from 'react-is';
import { UnknownProps } from './types';
import { doAsyncAct } from './internals/act';

export interface NodeLike {
  $$typeof: symbol | number;
  type?: NodeLike;
  props?: UnknownProps;
}

export interface ContextLike extends NodeLike {
  _context: {
    displayName?: string;
  };
}

export interface PortalLike extends NodeLike {
  containerInfo: HTMLElement;
}

/**
 * Return true if the value is either a React class or component declaration.
 */
export function isReactComponent(value: unknown): value is React.ComponentType {
  return typeof value === 'function';
}

/**
 * Return true if the value is a React class component instance.
 */
export function isReactComponentInstance(value: unknown): value is Function {
  return (
    typeof value === 'object' && !!value && 'isReactComponent' in value && 'constructor' in value
  );
}

/**
 * Return true if the value is structurally similar to a React element or node,
 * by checking for the existance of `$$typeof`.
 */
export function isReactNodeLike(value: unknown): value is NodeLike {
  return typeof value === 'object' && !!value && '$$typeof' in value;
}

/**
 * Return the `displayName` of a React context node, otherwise return "Context".
 */
export function getContextName(node: NodeLike): string {
  if (node.type) {
    return getContextName(node.type);
  }

  return ('_context' in node && (node as ContextLike)._context.displayName) || 'Context';
}

/**
 * Return the name of a React component, element, or node, by inferring metadata from
 * `$$typeof`. Utilizes the `react-is` package for common scenarios, but also takes
 * into account fibers and nodes not wrapped with React elements.
 */
export function getTypeName(type: unknown): string {
  if (!type) {
    return 'UNKNOWN';
  }

  if (isReactComponent(type)) {
    return type.displayName || type.name;
  }

  if (isReactComponentInstance(type)) {
    return getTypeName(type.constructor);
  }

  if (!isReactNodeLike(type) || typeof type === 'string') {
    return String(type);
  }

  const typeOf = type.$$typeof;

  if (ReactIs.isContextConsumer(type) || typeOf === ReactIs.ContextConsumer) {
    return `${getContextName(type)}.Consumer`;
  }

  if (ReactIs.isContextProvider(type) || typeOf === ReactIs.ContextProvider) {
    return `${getContextName(type)}.Provider`;
  }

  if (ReactIs.isForwardRef(type) || typeOf === ReactIs.ForwardRef) {
    return 'ForwardRef'; // We lose the component name
  }

  if (ReactIs.isFragment(type) || typeOf === ReactIs.Fragment) {
    return 'Fragment';
  }

  if (ReactIs.isLazy(type) || typeOf === ReactIs.Lazy) {
    return 'Lazy';
  }

  if (ReactIs.isMemo(type) || typeOf === ReactIs.Memo) {
    return `Memo(${getTypeName(type.type)})`;
  }

  if (ReactIs.isProfiler(type) || typeOf === ReactIs.Profiler) {
    return `Profiler(${type.props!.id})`;
  }

  if (ReactIs.isPortal(type) || typeOf === ReactIs.Portal) {
    const portal = type as PortalLike;

    return `Portal(${portal.containerInfo.id || portal.containerInfo.tagName.toLowerCase()})`;
  }

  if (ReactIs.isStrictMode(type) || typeOf === ReactIs.StrictMode) {
    return 'StrictMode';
  }

  if (ReactIs.isSuspense(type) || typeOf === ReactIs.Suspense) {
    return 'Suspense';
  }

  if (ReactIs.isElement(type)) {
    return getTypeName(type.type);
  }

  return String(type);
}

/**
 * Format a value for use within error messages.
 */
export function formatValue(value: unknown): string {
  const typeOf = typeof value;

  if (typeOf === 'string') {
    return `"${value}"`;
  } else if (typeOf === 'number' || typeOf === 'boolean' || value === null || value === undefined) {
    return `\`${String(value)}\``;
  }

  return `\`${getTypeName(value)}\``;
}

/**
 * Wait for async calls to finish by delaying execution.
 */
export async function wait(delay: number = 10): Promise<void> {
  await doAsyncAct(async () => {
    await new Promise(resolve => {
      setTimeout(resolve, delay);
    });
  });
}
