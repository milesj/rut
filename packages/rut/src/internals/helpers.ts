import React from 'react';
import Element from '../Element';
import { isReactNodeLike, NodeLike } from '../helpers';
import { UnknownProps } from '../types';

export function checkIsRutElement(value: unknown) {
  if (value instanceof Element) {
    return;
  }

  if (
    typeof value === 'object' &&
    value !== null &&
    (value.constructor.name === 'Element' || (value as Element).isRutElement === true)
  ) {
    return;
  }

  throw new Error('Expected a Rut `Element`.');
}

export function getPropFromElement<P, K extends keyof P>(
  element: Element<P>,
  name: K,
): P[K] | undefined {
  // @ts-ignore Allow internal access
  return element.element.props[name];
}

export function getPropForEmitting<P, K extends keyof P>(element: Element<P>, name: K): P[K] {
  let eventName = String(name);

  if (!eventName.startsWith('on')) {
    eventName = `on${eventName.charAt(0).toUpperCase()}${eventName.slice(1)}`;
  }

  const prop = getPropFromElement(element, eventName as K);

  if (!prop) {
    throw new Error(`Prop \`${eventName}\` does not exist.`);
  } else if (typeof prop !== 'function') {
    throw new TypeError(`Prop \`${eventName}\` is not a function.`);
    // @ts-ignore Allow internal access
  } else if (typeof element.element.type !== 'string') {
    throw new TypeError('Emitting events is only allowed on host components (DOM elements).');
  }

  return prop;
}

// Keep shallow equal in sync with React core!
// https://github.com/facebook/react/blob/master/packages/shared/shallowEqual.js
export function shallowEqual(objA: unknown, objB: unknown): boolean {
  if (Object.is(objA, objB)) {
    return true;
  }

  if (typeof objA !== 'object' || objA === null || typeof objB !== 'object' || objB === null) {
    return false;
  }

  const keysA = Object.keys(objA);
  const keysB = Object.keys(objB);

  if (keysA.length !== keysB.length) {
    return false;
  }

  // Test for A's keys different from B.
  // eslint-disable-next-line unicorn/no-for-loop
  for (let i = 0; i < keysA.length; i += 1) {
    if (
      !Object.prototype.hasOwnProperty.call(objB, keysA[i]) ||
      // @ts-ignore
      !Object.is(objA[keysA[i]], objB[keysA[i]])
    ) {
      return false;
    }
  }

  return true;
}

export function containsProps(props: UnknownProps, contains: UnknownProps): boolean {
  return Object.keys(contains).every(prop => props[prop] === contains[prop]);
}

/**
 * Memo and other exotics do not appear in the reconciled tree,
 * so we need to remove it and dig deeper to find the true element type.
 */
export function unwrapExoticType(element: NodeLike): React.ReactElement['type'] {
  const rootType = element.type!;

  if (typeof rootType === 'string' || typeof rootType === 'function') {
    return rootType;
  }

  if (isReactNodeLike(rootType)) {
    return unwrapExoticType(rootType);
  }

  return rootType;
}
