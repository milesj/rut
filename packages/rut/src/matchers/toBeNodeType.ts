import Element from '../Element';
import { MatchResult, NodeType } from '../types';
import { isRutElement } from '../internals/utils';
import { formatValue } from '../helpers';

// Keep in sync with React upstream!
// https://github.com/facebook/react/blob/master/packages/shared/ReactWorkTags.js
const nodeTypeMap: { [K in NodeType]: number | number[] } = {
  'class-component': 1,
  'forward-ref': 11,
  'function-component': 0,
  'host-component': 5,
  memo: [14, 15],
};

/**
 * Check that an element is a specific type of React node.
 * React nodes are based on React fiber work tags.
 */
export default function toBeNodeType(element: Element, type: NodeType): MatchResult {
  isRutElement(element);

  // @ts-ignore Allow internal access
  const fiberTag = element.element._fiber.tag;
  const nodeToTag = nodeTypeMap[type];

  if (nodeToTag === undefined) {
    throw new Error(
      `Invalid node type "${type}", expected one of: ${Object.keys(nodeTypeMap).join(', ')}`,
    );
  }

  const actualTypeName = formatValue(fiberTag);
  const typeName = formatValue(type);

  return {
    actual: actualTypeName,
    expected: typeName,
    message: `expected {{received}} to be a {{expected}}, found {{actual}}`,
    name: 'toBeNodeType',
    notMessage: `expected {{received}} not to be a {{expected}}, found {{actual}}`,
    passed: Array.isArray(nodeToTag) ? nodeToTag.includes(fiberTag) : nodeToTag === fiberTag,
    received: element.toString(),
  };
}
