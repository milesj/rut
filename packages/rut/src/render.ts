import React from 'react';
import AsyncResult from './AsyncResult';
import SyncResult from './SyncResult';
import { globalOptions } from './configure';
import { doAct, doAsyncAct } from './internals/act';
import { RendererOptions } from './types';

// export function render<Props extends object = {}>(
//   element: React.ReactElement<Props>,
//   options?: RendererOptions,
// ): SyncResult<Props> {
//   const opts = {
//     ...globalOptions,
//     ...options,
//   };

//   return doAct(() => new SyncResult(element, opts));
// }

export async function renderAndWait<Props extends object = {}>(
  element: React.ReactElement<Props>,
  options?: RendererOptions,
): Promise<AsyncResult<Props>> {
  const opts = {
    ...globalOptions,
    ...options,
  };

  return doAsyncAct(() => new AsyncResult(element, opts), opts.asyncMode);
}

export const render = renderAndWait;
