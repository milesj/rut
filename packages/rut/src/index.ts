/**
 * @copyright   2019, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import { act } from 'react-test-renderer';
import matchers from './matchers';
import { render, renderAndWait } from './render';
import mockSyntheticEvent from './mocks/event';

export * from './helpers';
export * from './types';

export { act, matchers, mockSyntheticEvent, render, renderAndWait };
