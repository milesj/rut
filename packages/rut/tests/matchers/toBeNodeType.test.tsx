/* eslint-disable rut/no-internals */

import React from 'react';
import { render } from '../../src/render';
import { NodeType } from '../../src/types';
import toBeNodeType from '../../src/matchers/toBeNodeType';
import { runMatcher } from '../helpers';
import { FuncComp, ClassComp, ForwardRefComp, MemoComp } from '../fixtures';
import Element from '../../src/Element';

describe.skip('toBeNodeType()', () => {
  it('errors if a non-Element is passed', () => {
    expect(() => {
      expect(123).toBeNodeType('function-component');
    }).toThrowError('Expected a Rut `Element`.');
  });

  it('errors if an invalid node type', () => {
    expect(async () => {
      expect((await render(<div />)).root).toBeNodeType(
        // @ts-ignore Allow invalid
        'unknown-type',
      );
    }).toThrowErrorMatchingSnapshot();

    expect(async () => {
      runMatcher(
        toBeNodeType(
          (await render(<div />)).root,
          // @ts-ignore Allow invalid
          'unknown-type',
        ),
      );
    }).toThrowErrorMatchingSnapshot();
  });

  const nodeTypes = {
    'class-component': <ClassComp />,
    'forward-ref': <ForwardRefComp />,
    'function-component': <FuncComp />,
    'host-component': <div />,
    memo: <MemoComp />,
  };

  // eslint-disable-next-line
  Object.entries(nodeTypes).forEach(async ([testTypeName, testNode]) => {
    const expectedNode = new Element(
      // @ts-ignore
      (await render(<div>{testNode}</div>)).root.element.children[0],
    );
    const typeName = testTypeName as NodeType;

    if (typeof expectedNode === 'string') {
      return;
    }

    describe(`"${typeName}" type`, () => {
      it('passes when types match', () => {
        expect(() => {
          runMatcher(toBeNodeType(expectedNode, typeName));
        }).not.toThrowError();
      });

      it('errors when types dont match', () => {
        expect(() => {
          runMatcher(
            toBeNodeType(
              expectedNode,
              typeName === 'function-component' ? 'class-component' : 'function-component',
            ),
          );
        }).toThrowErrorMatchingSnapshot();
      });

      it('passes when types match (not negation)', () => {
        expect(() => {
          runMatcher(
            toBeNodeType(
              expectedNode,
              typeName === 'function-component' ? 'class-component' : 'function-component',
            ),
            true,
          );
        }).not.toThrowError();
      });

      it('errors when types dont match (not negation)', () => {
        expect(() => {
          runMatcher(toBeNodeType(expectedNode, typeName), true);
        }).toThrowErrorMatchingSnapshot();
      });
    });
  });
});
