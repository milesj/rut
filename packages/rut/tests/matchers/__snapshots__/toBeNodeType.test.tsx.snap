// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toBeNodeType() "class-component" type errors when types dont match (not negation) 1`] = `"expected <ClassComp /> not to be a \\"class-component\\", found \\"class-component\\""`;

exports[`toBeNodeType() "class-component" type errors when types dont match 1`] = `"expected <ClassComp /> to be a \\"function-component\\", found \\"class-component\\""`;

exports[`toBeNodeType() "forward-ref" type errors when types dont match (not negation) 1`] = `"expected <ForwardRef /> not to be a \\"forward-ref\\", found \\"forward-ref\\""`;

exports[`toBeNodeType() "forward-ref" type errors when types dont match 1`] = `"expected <ForwardRef /> to be a \\"function-component\\", found \\"forward-ref\\""`;

exports[`toBeNodeType() "function-component" type errors when types dont match (not negation) 1`] = `"expected <FuncComp /> not to be a \\"function-component\\", found \\"function-component\\""`;

exports[`toBeNodeType() "function-component" type errors when types dont match 1`] = `"expected <FuncComp /> to be a \\"class-component\\", found \\"function-component\\""`;

exports[`toBeNodeType() "host-component" type errors when types dont match (not negation) 1`] = `"expected <div /> not to be a \\"host-component\\", found \\"host-component\\""`;

exports[`toBeNodeType() "host-component" type errors when types dont match 1`] = `"expected <div /> to be a \\"function-component\\", found \\"host-component\\""`;

exports[`toBeNodeType() "memo" type errors when types dont match (not negation) 1`] = `"expected <Memo(FuncComp) /> not to be a \\"memo\\", found \\"memo\\""`;

exports[`toBeNodeType() "memo" type errors when types dont match 1`] = `"expected <Memo(FuncComp) /> to be a \\"function-component\\", found \\"memo\\""`;

exports[`toBeNodeType() errors if an invalid node type 1`] = `"Invalid node type \\"unknown-type\\", expected one of: class-component, forward-ref, function-component, host-component, memo"`;

exports[`toBeNodeType() errors if an invalid node type 2`] = `"Invalid node type \\"unknown-type\\", expected one of: class-component, forward-ref, function-component, host-component, memo"`;
